### PR TITLE
add Accumulator AndNot, OrAcc/AndNotAcc merge, per-call concurrency, and pooling config

### PR DESCRIPTION
**Pull request by aliszka**
## Motivation

The Accumulator ships with `Or` only. This adds subtraction (`AndNot`), accumulator-to-accumulator merging (`OrAcc`/`AndNotAcc`) — so partial accumulators built on separate goroutines can be folded into one — per-call concurrency (`WithConc`), and configurable block retention (`WithRetainedBlocks`) so a pooled accumulator can reuse its staging memory across resets.

## Approach

`AndNot` bit-clears into existing staging blocks and never creates one, so `Or` and `AndNot` interleave freely and the result reflects every call in order; the key walk gallops over whichever side is much larger. Emptied ranges keep their 8KB blocks until `Reset` but never cost result bytes.

`OrAcc`/`AndNotAcc` treat another accumulator's staged state as the source. The receiver is the destination; `other` is only read and must stay unmodified during the call — blocks are copied or word-merged, never adopted. `OrAcc` on itself is a no-op, `AndNotAcc` on itself empties every range.

`WithConc` caps in-call fan-out: `<= 1` keeps every call serial and unbounded parallelism is requested with a large value (e.g. `math.MaxInt`), so the zero Accumulator stays serial with no sentinel (`concCap` resolves `max(1, n)`). This intentionally diverges from `Bitmap`'s released `*Conc` methods (`<= 0` uncapped) — the Accumulator is a new, separately documented entity, and its two config knobs share one rule instead: a small value means "this knob's default", a large value the extreme. The cap survives `Reset`. Concurrent paths need no locks: the calling goroutine pre-creates every missing block, so workers only read the index and write disjoint blocks via galloping `searchFrom`. The Accumulator remains not safe for concurrent use — parallelism is internal to each call.

`WithRetainedBlocks(n)` sizes how many staging blocks `Reset` keeps as spares for the next union (each ~8KB, one per touched 64K range). `<= 0` retains nothing — the default, so pooled reuse is opt-in — and a large value retains everything a union touched; the setting survives `Reset`, so a pool configures it once and checks the accumulator in and out repeatedly. This replaces the previous hardcoded 16-block spare cap, and the keys/blocks slot-retention floor lifts to `max(1024, retained)` so a large setting keeps its backing slices. A warm accumulator with retention configured, building into a pooled result buffer, allocates nothing across `Reset`.

Deposits stay at one call per container: `stageBlock` resolves a block with a single local binary search (the generic `slices.BinarySearch` call alone cost double-digit percent on singleton folds), deposit switches live inline, and fresh blocks adopt dense payloads by plain `copy`. Benchmarked vs main: singleton folds within ~5%, large-dense 5–8% faster, wide par to 3% faster.

Also: CI runs the suite under `-race`; a dead loop-var copy in `concurrentlyInRanges` removed.

## Key areas for review

- The no-locks argument — pre-create before fan-out plus unique source keys ⇒ disjoint worker writes; nothing on the worker path touches `lastIdx`.
- The `|=` deposit switch exists twice (serial `Or` and `orRange`) — deliberate: a shared helper measured +10.7% on singleton folds.
- `andNotRange`/`andNotAccRange` partitioning — workers re-seed the source cursor by binary search and clamp both walks to their slice.
- The two config knobs' shared `small → default, large → extreme` rule, and its deliberate divergence from `Bitmap.*Conc` (documented on `WithConc`).

## Risks

- A fan-out race would be silent corruption — mitigated by the pre-create design and CI `-race` over every table in capped and uncapped variants, including concurrent self-merge.
- `WithConc` could be misread as thread-safety; the type and method docs say otherwise.
- `WithConc`'s `<= 1`-serial convention differs from the released `Bitmap.*Conc` `<= 0`-unlimited; called out on `WithConc` so a `0` argument isn't misread. `Bitmap.*Conc` is intentionally left unchanged.
- Subtraction never shrinks staging (documented); pooled retention is bounded by `WithRetainedBlocks`, which defaults to 0 (retain nothing).

## Testing

Every table runs under serial / `WithConc(4)` / `WithConc(math.MaxInt)` constructors, each case driven twice through a shared `accTarget` surface — map-based reference vs accumulator. Covers overlap/disjoint/subset ranges, gallop skew, the array/bitmap cutoff, emptied ranges both directions, self-merge, nil/empty sources, interleaved ops, block-creation invariants, copy-length pinning, unknown-type panics, mid-build guards for all four deposit calls, `Reset` retuning, block retention (default-nothing, explicit-zero, negative-clamp, retain-to-cap-and-reuse, survives-`Reset`), `concCap` resolution across the value range, and a post-subtraction serialization round-trip. Full suite green under `-race`.

### CODE REVIEW

File: accumulator.go [hunk 1]
```diff
@@ -113,12 +233,141 @@ func (acc *Accumulator) Or(bm *Bitmap) {
 	}
 	acc.gen++
 	n := bm.keys.numKeys()
+	conc := calcConcurrency(n, minContainersPerRoutine, acc.concCap())
+	if conc <= 1 {
```
Comment (amourao) [inline accumulator.go:237]: **issue:** FromBuffer is the public producer, and WithConc(...).Or reaches the worker path.
The no-lock argument at accumulator.go:272 assumes unique source keys, but FromBuffer does not validate that invariant. A race probe placed duplicate keys across two partitions. The race detector reported both workers updating the same destination word in orRange.
Please validate strict ordering before starting workers, or fall back to the serial path when keys repeat. Add a race test with the duplicate across a partition boundary.

File: accumulator.go [hunk 2]
```diff
@@ -279,39 +680,43 @@ func (acc *Accumulator) buildInto(ra *Bitmap, cards []int) {
...
 			break
 		}
-		clear(b)
 		acc.free = append(acc.free, b)
 	}
-	if cap(acc.keys) > maxRetainedSlots {
+	// Keep the keys/blocks backing arrays if their capacity is within the
+	// slot floor or what the retained blocks need; drop an abnormally large
+	// one so a single wide union can't pin it for the accumulator's life.
+	if cap(acc.keys) > max(maxRetainedSlots, acc.maxRetained) {
```
Comment (amourao) [inline accumulator.go:719]: **issue:** A caller reaches this with WithRetainedBlocks(n), a union over n > 1024 ranges, and Reset.
At accumulator.go:719, the code compares allocator-grown capacity with the configured block count. With n=1025, the key and block capacities grew to 1,280 and 1,135. Reset then set both slices to nil, and the next pooled cycle performed 24 allocations.
Please base the drop decision on the touched count, or resize once to a retained capacity instead of discarding the slices. Add an allocation test where the configured count equals the touched range count.

File: accumulator.go [hunk 3]
```diff
@@ -102,7 +185,44 @@ func (acc *Accumulator) block(key uint64) []uint16 {
...
 	acc.lastIdx = lo
-	return b
+	return lo
+}
+
+// searchFrom returns the smallest index in (from, to) at which acc.keys[i]
+// >= k, or to if there is none: exponential expansion to bracket k, then
+// binary search — O(log gap). from may be -1 to include index 0 in the
+// search.
+func (acc *Accumulator) searchFrom(from, to int, k uint64) int {
```
Comment (amourao) [inline accumulator.go:195]: **nitpick:** accumulator.go:195, keys.go:113, and bitmap_opt.go:2118 perform the same expansion, clamp, and binary-search steps over different key accessors.
Please use a benchmarked shared lower-bound kernel with representation adapters. If an adapter costs too much on this path, record the measurement that justifies keeping the copies and add direct boundary tests for each one.

Comment (aliszka) [reply accumulator.go:195]: Tried two ways to share the loop and benchmarked both against the current copies. Each figure is the fastest of six runs, A/B'd in a single binary:

| approach | node | acc | masked |
|---|---|---|---|
| generics + adapter types | +202% | +149% | +168% |
| cursor (caller keeps the fetch) | +299% | +246% | +268% |

End-to-end it's mostly amortized away, except where key-walking dominates: `Intersects` goes 23ns → 43ns (+88%) with the generic version.

Same cause both times — sharing this loop at runtime puts an indirection inside a body that is one comparison and a shift. Generics route type-parameter method calls through a dictionary, and since `node` is `[]uint64` the node and accumulator instantiations share a GC shape, collapse into one body, and can't specialize. A cursor's `probe`/`observe` exceed the inline budget, so it's two real calls per probe.

I didn't benchmark a stdlib kernel because it can't work here: a `node` interleaves keys with offsets, so there's no slice for `sort.Search`/`slices.BinarySearchFunc` to address, and both are plain bisect where these are O(log gap). Codegen would be free at runtime but seems disproportionate for 3×20 lines.

So keeping the copies. Taking the second half though — `Accumulator.searchFrom` and `searchKeyedMaskedFrom` have no direct tests today (`node.searchFrom` has `TestSearchFrom`), and that's worth fixing on its own: the three aren't quite the same function. Duplicate masked keys mean that variant needs first-occurrence semantics, while the other two can early-return on an exact match because their keys are unique.

File: accumulator.go (hunk 1 above)
Comment (aliszka) [reply accumulator.go]: Real, but it takes a bitmap with repeated keys, which only `FromBuffer` over corrupt bytes can produce — and `FromBuffer` validates no offsets, sizes or type tags either, so guarding this one invariant wouldn't make a malformed bitmap safe. Worth knowing the shipped `Bitmap.OrConc` is worse on the same input: `containerOrAlt` merges in place, so it races *and* returns a wrong cardinality.

File: accumulator.go (hunk 2 above)
Comment (aliszka) [reply accumulator.go]: Fixed: keys/blocks/free now follow the block budget at 3× it, which clears append's overshoot so a union of exactly the configured size keeps its slices instead of rebuilding them every cycle. `free`'s own array was bounded by nothing before — it survived even `WithRetainedBlocks(0)` — and is released now too; allocation test added for the configured-size case.

**Review by antas-marcin — APPROVED** [review]
